### PR TITLE
[extractor/common] Add `default` parameter to `_search_json`

### PR DESCRIPTION
**Pull request by coletdjnz**
### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [X] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [X] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

---

### Description of your *pull request* and other information

Explanation of your *pull request* in arbitrary form goes here. Please **make sure the description explains the purpose and effect** of your *pull request* and is worded well enough to be understood. Provide as much **context and examples** as possible.

It is expected YoutubeWebArchive to fail to extract the json for some pages, and it should not raise a warning.

### CODE REVIEW

File: yt_dlp/extractor/archiveorg.py [hunk 1]
```diff
@@ -486,9 +486,9 @@ def _extract_metadata(self, video_id, webpage):
         search_meta = ((lambda x: self._html_search_meta(x, webpage, default=None)) if webpage else (lambda x: None))
         player_response = self._search_json(
             self._YT_INITIAL_PLAYER_RESPONSE_RE, webpage, 'initial player response',
-            video_id, fatal=False)
+            video_id, fatal=False, default={})
         initial_data = self._search_json(
-            self._YT_INITIAL_DATA_RE, webpage, 'initial data', video_id, fatal=False)
+            self._YT_INITIAL_DATA_RE, webpage, 'initial data', video_id, fatal=False, default={})
```
Comment (pukkandan) [inline yt_dlp/extractor/archiveorg.py:491]: default implies non-fatal

File: yt_dlp/extractor/common.py [hunk 2]
```diff
@@ -1188,13 +1188,16 @@ def _search_regex(self, pattern, string, name, default=NO_DEFAULT, fatal=True, f
...
         # NB: end_pattern is only used to reduce the size of the initial match
         return self._parse_json(
-            self._search_regex(rf'{start_pattern}\s*(?P<json>{{{contains_pattern}}})\s*{end_pattern}',
-                               string, name, group='json', fatal=fatal) or '{}',
-            video_id, fatal=fatal, ignore_extra=True, **kwargs) or {}
+            self._search_regex(
+                rf'{start_pattern}\s*(?P<json>{{{contains_pattern}}})\s*{end_pattern}',
+                string, name, group='json', fatal=fatal,
+                default='{}' if default is not NO_DEFAULT else NO_DEFAULT),
+            video_id, fatal=fatal, ignore_extra=True, **kwargs) or {} if default is NO_DEFAULT else default
```
Comment (pukkandan) [inline yt_dlp/extractor/common.py:1200]: I don't think this behaves correctly. Pls double check

What we want:
* `fatal=True` => throw regex error
* `fatal=False` => print regex warning, but not json warning and return `{}`
* `default=xxx` => print neither warning and return `xxx` (not `json.loads(xxx)`)

Note that `fatal` should be ignored in the last case

(split into 2 statements if needed)

You can merge once you confirm this behavior

Comment (coletdjnz) [reply yt_dlp/extractor/common.py]: hmm ok. I've had a go at fixing it up, you might want to have a look to see if you can simplify it.

would add a test but not sure if we have an easy way for testing warnings

**Comment by pukkandan** [comment]
It ended up being pretty complicated 😅

**Pull request merged**